File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        // Iterative
        // base case
        if (preorder.empty()) return nullptr;
        
        TreeNode* root = new TreeNode(preorder[0]);
        int pre = 1, ino = 0;
        stack<TreeNode*> stk;
        stk.push(root);
        
        while (pre < preorder.size()){
            TreeNode* curr = new TreeNode(preorder[pre++]), *prev = nullptr;
            
            while (!stk.empty() && stk.top() -> val == inorder[ino]){
                prev = stk.top();
                stk.pop();
                ++ino;
            }
            
            if (prev)
                prev -> right = curr;
            else
                stk.top() -> left = curr;
            stk.push(curr);
        }
        
        return root;
    }
};

File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        # Iterative
        if not preorder:
            return None
        
        root = TreeNode(preorder[0])
        pre, ino = 1, 0
        stack = []
        stack.append(root)
        
        while pre < len(preorder):
            curr = TreeNode(preorder[pre])
            pre += 1
            prev = None
            
            while stack and stack[-1].val == inorder[ino]:
                prev = stack.pop()
                ino += 1
            
            if prev:
                prev.right = curr
            else:
                stack[-1].left = curr
        
            stack.append(curr)
            
        return root